scipy/signal/_peak_finding.py

```diff
@@ -475,97 +482,59 @@ def peak_widths(x, peaks, rel_height=0.5, prominence_data=None, wlen=None):
     >>> from scipy.signal import chirp, find_peaks, peak_widths
     >>> import matplotlib.pyplot as plt
 
-    Create a test signal with growing peak widths
+    Create a test signal with two overlayed harmonics
 
-    >>> x = np.linspace(0, 500, 500)
-    >>> x = abs(chirp(x, 1e-4, x.max(), 1.1e-2)) + 2.0 * x / x.max()
+    >>> x = np.linspace(0, 6 * np.pi, 1000)
+    >>> x = np.sin(x) + 0.6 * np.sin(2.6 * x)
 
     Find all peaks and calculate their widths at the relative height of 0.5
+    (contour line at half the prominence height) and 1 (at the lowest contour
+    line at full prominence height).
 
     >>> peaks, _ = find_peaks(x)
-    >>> widths, heights, lpos, rpos = peak_widths(x, peaks, rel_height=0.5)
-    >>> widths
-    array([77.7462348 , 62.19574776, 45.57709222, 37.902356  , 33.33210357,
-           29.81097122])
+    >>> results_half = peak_widths(x, peaks, rel_height=0.5)
+    >>> results_half[0]  # widths
+    array([ 64.25172825,  41.29465463,  35.46943289, 104.71586081,
+            35.46729324,  41.30429622, 181.93835853,  45.37078546])
+    >>> results_full = peak_widths(x, peaks, rel_height=1)
+    >>> results_full[0]  # widths
+    array([181.9396084 ,  72.99284945,  61.28657872, 373.84622694,
+        61.78404617,  72.48822812, 253.09161876,  79.36860878])
 
     Plot signal, peaks and contour lines at which the widths where calculated
 
     >>> plt.plot(x)
-    >>> plt.plot(peaks, x[peaks], "x", color="C1")
-    >>> plt.hlines(y=heights, xmin=lpos, xmax=rpos, color="C1")
+    >>> plt.plot(peaks, x[peaks], "x")
+    >>> plt.hlines(*results_half[1:], color="C2")
+    >>> plt.hlines(*results_full[1:], color="C3")
     >>> plt.show()
     """
-    x = np.asarray(x)
-    peaks = np.asarray(peaks)
-
-    if peaks.size == 0:
-        # Handle empty peaks
-        return tuple(np.array([]) for _ in range(4))
-
+    # Inner function expects `x` to be C-contiguous
+    x = np.asarray(x, order='C', dtype=np.float64)
     if x.ndim != 1:
         raise ValueError('`x` must have exactly one dimension')
+
+    peaks = np.asarray(peaks)
+    if peaks.size == 0:
+        # Empty arrays default to np.float64 but are valid input
+        peaks = np.array([], dtype=np.intp)
+    try:
+        # Safely convert to C-contiguous array of type np.intp
+        peaks = peaks.astype(np.intp, order='C', casting='safe',
+                             subok=False, copy=False)
+    except TypeError:
+        raise TypeError("Cannot safely cast `peaks` to dtype('intp')")
```

Comment:
Because with  `astype` I can use the `safe` casting rule which means input arrays of type float don't get implicitly cast to integers. I wanted the function visibly fail for inputs that don't make sense like supplying an array of floats as indices.